tests/testdeprecations.py

```diff
@@ -19,40 +19,188 @@ def specific_filter():
-        param.as_unicode(1)
+        p = P()
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            p.n = 1
 
 
-def test_deprecate_is_ordered_dict():
-    with pytest.raises(param._utils.ParamDeprecationWarning):
-        param.is_ordered_dict(dict())
+class TestDeprecateInitModule:
 
+    def test_deprecate_as_unicode(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.as_unicode(1)
 
-def test_deprecate_Number_set_hook():
-    class P(param.Parameterized):
-        n = param.Number(set_hook=lambda obj, val: val)
+    def test_deprecate_is_ordered_dict(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.is_ordered_dict(dict())
 
-    p = P()
-    with pytest.raises(param._utils.ParamDeprecationWarning):
-        p.n = 1
+    def test_deprecate_produce_value(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.produce_value(1)
 
+    def test_deprecate_hasbable(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.hashable('s')
 
-def test_deprecate_overridable_property():
-    with pytest.raises(param._utils.ParamDeprecationWarning):
-        class A:
-            def __init__(self):
-                self._x = 1
-            def _get_x(self):
-                return self._x
-            x = param.parameterized.overridable_property(_get_x)
+    def test_deprecate_named_objs(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.named_objs(dict())
+
+    def test_deprecate_normalize_path(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.normalize_path()
+
+    def test_deprecate_abbreviate_path(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.abbreviate_paths()
+
+
+class TestDeprecateParameterizedModule:
+
+    def test_deprecate_overridable_property(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            class Foo:
+                def _x(self): pass
+                x = param.parameterized.overridable_property(_x)
+
+    def test_deprecate_batch_watch(self):
+        class P(param.Parameterized):
+            x = param.Parameter()
+
+        p = P()
+
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            with param.parameterized.batch_watch(p):
+                pass
+
+    def test_deprecate_add_metaclass(self):
+        class MC(type): pass
+
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            @param.parameterized.add_metaclass(MC)
+            class Base: pass
+
+    def test_deprecate_recursive_repr(self):
+        with pytest.raises(param._utils.ParamDeprecationWarning):
+            param.parameterized.recursive_repr(lambda: '')
+
+
+class TestDeprecateParameters:
+
+    def test_deprecate_print_param_defaults(self):
```

Comment:
I would maybe take advantage of using a class and only create the P instance once with `setup_method` https://docs.pytest.org/en/latest/how-to/xunit_setup.html.

Reply:
Yep true the tests I write for Param aren't usually very DRY. I think it's because I've struggled with the original test suites, where some test modules have a setup that defines a Parameterized class with lots of Parameters, and when working in these modules I end up spending way too much time scrolling 🙃 So I've come down to preferring repeating myself and having self-contained tests.